tests/integration/kernel/src/main.c

```diff
@@ -0,0 +1,291 @@
+/* Thread B gives the semaphore for Thread A to continue. */
+/* Once the control is returned back to Thread A, it extracts all items from the stack. */
+/* Verify the data's correctness. */
+
+
+#define STACK_LEN	8
+#define MAX_ITEMS	8
+
+K_STACK_DEFINE(stack, STACK_LEN);
+stack_data_t stack_data[MAX_ITEMS];
+
+
+static void thread_entry_fn_stack(void *p1, void *p2, void *p3)
+{
+	stack_data_t *rx_data;
+
+	/* read data in a stack */
+	for (int i = 0; i < STACK_LEN; i++) {
+		k_stack_pop((struct k_stack *)p1, (stack_data_t *)&rx_data, K_NO_WAIT);
+	}
+
+	/* Push data into a stack */
+	for (int i = 0; i < STACK_LEN; i++) {
+		stack_data[i] *= i;
+		k_stack_push((struct k_stack *)p1, (stack_data_t)&stack_data[i]);
+	}
+
+	/* Give control back to Thread A */
+	k_sem_give(&sema);
+}
+
+
+ZTEST(kernel, test_stack_usage)
+{
+	stack_data_t *rx_data;
+
+	/* Init binary semaphore */
+	k_sem_init(&sema, 0, 1);
+
+	/* Push data into a stack */
+	for (int i = 0; i < STACK_LEN; i++) {
+		stack_data[i] = i;
+		k_stack_push(&stack, (stack_data_t)&stack_data[i]);
+	}
+
+	/* Create the Thread B */
+	k_tid_t tid = k_thread_create(&tdata, tstack, STACK_SIZE,
+					thread_entry_fn_stack, &stack, NULL, NULL,
+					K_PRIO_PREEMPT(0), K_INHERIT_PERMS, K_NO_WAIT);
+
+	/* Let the thread B run */
+	k_sem_take(&sema, K_FOREVER);
+
+	/* read data in a stack */
+	for (int i = 0; i < STACK_LEN; i++) {
+		k_stack_pop(&stack, (stack_data_t *)&rx_data, K_NO_WAIT);
+	}
+
+	/* Clear the spawn thread */
+	k_thread_abort(tid);
+}
+
+/* MUTEX SCENARIO */
+/* Initialize the mutex. */
+/* Start the two Threads with the entry points. */
+/* The entry points change variable and use lock and unlock mutex functions to */
+/* synchronize the access to that variable. */
+/* Join all threads. */
+/* Verify the variable. */
+
+#define NUMBER_OF_ITERATIONS 1000
+
+static uint32_t mutex_data;
+static struct k_thread tdata_2;
+static K_THREAD_STACK_DEFINE(tstack_2, STACK_SIZE);
+
+K_MUTEX_DEFINE(mutex);
+
+static void thread_entry_fn_mutex(void *p1, void *p2, void *p3)
+{
+	uint32_t i;
+
+	k_mutex_lock((struct k_mutex *)p1, K_FOREVER);
+	for (i = 0; i < NUMBER_OF_ITERATIONS; i++) {
+		mutex_data += 1;
+	}
+	k_mutex_unlock((struct k_mutex *)p1);
+}
+
+static void thread_entry_fn_mutex_2(void *p1, void *p2, void *p3)
+{
+	uint32_t i;
+
+	k_mutex_lock((struct k_mutex *)p1, K_FOREVER);
+	for (i = 0; i < NUMBER_OF_ITERATIONS; i++) {
+		mutex_data -= 1;
+	}
+	k_mutex_unlock((struct k_mutex *)p1);
+}
+
+ZTEST(kernel, test_mutex_usage)
+{
+
+    /* Create the Thread A */
+	k_tid_t tid = k_thread_create(&tdata, tstack, STACK_SIZE,
+					thread_entry_fn_mutex, &mutex, &mutex_data, NULL,
+					K_PRIO_PREEMPT(0), K_INHERIT_PERMS, K_NO_WAIT);
+
+	/* Create the Thread B */
+	k_tid_t tid2 = k_thread_create(&tdata_2, tstack_2, STACK_SIZE,
+					thread_entry_fn_mutex_2, &mutex, &mutex_data, NULL,
+					K_PRIO_PREEMPT(0), K_INHERIT_PERMS, K_NO_WAIT);
```

Comment:
A couple of things ...

1. `&mutex_data is passed as a parameter to the spawned threads, but the spawned threads do not use that parameter. They use the global variable.
2. Although I think that the threads are synchronized, it is not the mutex that is synchronizing them, but rather the scheduler that is doing so.

Reply:
You only can check mutex_data is 0, the middle step is unknown.